graspologic/layouts/auto.py

```diff
@@ -0,0 +1,347 @@
+
+    Parameters
+    ----------
+    graph : :class:`networkx.Graph`
+        The graph to generate a layout for. This graph may have edges pruned if the
+        count is too high and only the largest connected component will be used to
+        automatically generate a layout.
+    min_dist : float
+        The effective minimum distance between embedded points. Default is ``0.75``.
+        Smaller values will result in a more clustered/clumped embedding where nearby
+        points on the manifold are drawn closer together, while larger values will
+        result on a more even dispersal of points. The value should be set relative to
+        the ``spread`` value, which determines the scale at which embedded points will
+        be spread out.
+    n_neighbors : int
+        The size of local neighborhood (in terms of number of neighboring sample points)
+        used for manifold approximation. Default is ``25``. Larger values result in
+        more global views of the manifold, while smaller values result in more local
+        data being preserved.
+    max_edges : int
+        The maximum number of edges to use when generating the embedding.  Default is
+        ``10000000``. The edges with the lowest weights will be pruned until at most
+        ``max_edges`` exist. Warning: this pruning is approximate and more edges than
+        are necessary may be pruned. Running in 32 bit enviornment you will most
+        likely need to reduce this number or you will out of memory.
+
+    Returns
+    -------
+    Tuple[nx.Graph, List[NodePosition]]
+        The largest connected component and a list of NodePositions for each node in
+        the largest connected component. The NodePosition object contains:
+        - node_id
+        - x coordinate
+        - y coordinate
+        - size
+        - community
+
+    References
+    ----------
+    .. [1] McInnes, L, Healy, J, UMAP: Uniform Manifold Approximation and Projection
+        for Dimension Reduction, ArXiv e-prints 1802.03426, 2018
+    .. [2] Böhm, Jan Niklas; Berens, Philipp; Kobak, Dmitry. A Unifying Perspective
+        on Neighbor Embeddings along the Attraction-Repulsion Spectrum. ArXiv e-prints
+        2007.08902v1, 17 Jul 2020.
+    """
+
+    lcc_graph, tensors, labels = _node2vec_for_layout(graph, max_edges)
+    points = umap.UMAP(min_dist=min_dist, n_neighbors=n_neighbors).fit_transform(
+        tensors
+    )
+    positions = _node_positions_from(lcc_graph, labels, points)
+    return lcc_graph, positions
+
+
+def _largest_connected_component(graph: nx.Graph) -> nx.Graph:
+    largest_component = max(nx.connected_components(graph), key=len)
+    return graph.subgraph(largest_component).copy()
+
+
+def _approximate_prune(graph: nx.Graph, max_edges_to_keep: int = 1000000):
+    num_edges = len(graph.edges())
+    logger.info(f"num edges: {num_edges}")
+
+    if num_edges > max_edges_to_keep:
+        histogram, bins = histogram_edge_weight(graph, bin_directive=100)
```

Comment:
Precisely, and bin directive could absolutely be exposed - I just didn't to start with since I'm taking the "let's not make this even more complicated than it has to be" initial tack

Binary search wouldn't work because all you have in the histogram at each index is the count in that bin.  That's why we do this kinda janky approximate approach here - we're counting the population in each bucket, starting with the largest weights and moving down.  Eventually we'll find a bucket that would tip us over our threshold, and we prune the entire bucket and everything smaller than it. 

I thought about exposing this in `graph_cuts.py`, but I would want something to actually find me precisely `max_edges_to_keep` and make the cuts to get to that point, which would require us to identify the edges within the bin, sort them, and prune the smallest until we hit that threshold.  And that is totally reasonable, but there's already so much in this PR I wasn't too keen on trying to do it right now, and so we went with this brute force approximate approach to start with.